snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,282 @@
+        return result
+
+    def logout(self):
+        self.conf.clear()
+        self.conf.save()
+
+    def register_name(self, name):
+        data = dict(snap_name=name, series=DEFAULT_SERIES)
+        auth = _macaroon_auth(self.conf, 'package_upload')
+        response = self.post('register-name/',
+                             data=json.dumps(data),
+                             headers={'Authorization': auth,
+                                      'Content-Type': 'application/json'})
+        if not response.ok:
+            # if (response['errors'] == ['Authorization Required']
+            #     and (response.headers['WWW-Authenticate']
+            #          == "Macaroon needs_refresh=1":
+            # Refresh the discharge macaroon and retry
+            pass
+        return response
+
+    def upload(self, binary_path, snap_name):
+        if self.conf.get('package_upload') is None:
+            raise InvalidCredentialsError()
+        self.updown = requests.Session()
+        data = _upload.upload_files(binary_path, self.updown)
+        success = data.get('success', False)
+        if not success:
+            return data
+
+        result = _upload.upload_app(self, snap_name, data)
+        return result
+
+    def download(self, snap_name, channel, download_path, arch=None):
+        if arch is None:
+            arch = snapcraft.ProjectOptions().deb_arch
+        self.cpi = CPIClient(self.conf)
+        logger.info('Getting details for {}'.format(snap_name))
+        package = self.cpi.search_package(snap_name, channel, arch)
+        if package is None:
+            raise SnapNotFoundError(snap_name, channel, arch)
+        return self.download_snap(snap_name, channel, arch,
+                                  download_path, package['download_url'],
+                                  package['download_sha512'])
+
+    def get_macaroon(self, acl):
+        response = self.post('../../api/2.0/acl/{}/'.format(acl),
+                             data=json.dumps({}),
+                             headers={'Content-Type': 'application/json'})
+        if response.ok:
+            return response.json()['macaroon'], None
+        else:
+            return None, response.text
+
+    def get_discharges(self, email, password, one_time_password, macaroons):
+        data = dict(email=email, password=password,
+                    macaroons=[(k, v) for k, v in macaroons.items()])
+        if one_time_password:
+            data['otp'] = one_time_password
+        try:
+            response = self.sso.session.post(
+                '/tokens/discharge', data=data,
+                headers={'Content-Type': 'application/json',
+                         'Accept': 'application/json'})
+            return dict(response.content['discharge_macaroons']), None
+        except sso.ApiException as err:
+            return None, err.body
+        except sso.UnexpectedApiError as err:
+            return None, err.json_body
```

Comment:
This looks like go. I'm liking it a lot in go, but in python should we raise the exception instead? It's more common to see the caller doing try catch instead of if error.